### PR TITLE
py3k support, sanity check for libLLVM glob (was: Travis CI + test case)

### PR DESCRIPTION
**Pull request by androm3da**
This is a (mostly) unrelated PR from the previous one.  Doesn't include the verification, just a CI config and a simple test case from the README.

Only change to the normal functionality is a sanity check for when the glob comes up empty.  I added this because with only `llvm` (without `libllvm`), it fails mysteriously.

### CODE REVIEW

**Comment by androm3da** [comment]
Oops, it looks like there's a perfectly good test case that's here already.  I deleted the new test case.

Here's the (passing) test results for py2, py3 and pypy: https://travis-ci.org/androm3da/llvmcpy/builds/194156028

File: llvmcpy/llvm.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def header():
         zeroth_argument = ["self"] if is_class_method else []
         # Discard None arguments, they have been removed
         args = filter(lambda x: x is not None, function_arguments)
-        function_arguments_str = ", ".join(zeroth_argument + args)
+        function_arguments_str = ", ".join(zeroth_argument + list(args))
```
Comment (aleclearmind) [inline llvmcpy/llvm.py]: We can fix this separately.

File: README.rst [hunk 2]
```diff
@@ -26,9 +26,12 @@ versions of the `LLVM-C API`_ should be basically effortless.
 
 To try it out install LLVM and get `llvmcpy` using `pip`::
 
-    sudo apt-get install llvm
+    sudo apt-get install llvm libllvm3.8
     pip install llvmcpy
 
+See [apt.llvm.org](http://apt.llvm.org/) for more package sources
```
Comment (aleclearmind) [inline README.rst]: Beware: this is not Markdown, this is [reStructuredText](http://docutils.sourceforge.net/docs/user/rst/quickref.html).

**Comment by aleclearmind** [comment]
Testing various Python version is very useful, but I think it would be even more useful to test as many *LLVM* versions as possible.

On the official LLVM apt repository I can find the following versions:

http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.4/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.5/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.6/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.7/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.8/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-3.9/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-4.0/main/binary-amd64/Packages
http://apt.llvm.org/trusty/dists/llvm-toolchain-trusty-5.0/main/binary-amd64/Packages

Maybe we can have Travis test llvmcpy with all the versions of Python + the current stable LLVM (currently 3.9.1) and then another set of test with all the versions of LLVM available on apt.llvm.org and Python 2.7 or Python 3.x.

File: llvmcpy/llvm.py (hunk 1 above)
Comment (androm3da) [inline llvmcpy/llvm.py]: It was required to get any py3k tests to pass, that's why it was included here.

I can put it in a separate PR, I just figured it would make sense to have all the tests green.  If I remove it, should I disable the py3k tests?

File: README.rst (hunk 2 above)
Comment (androm3da) [inline README.rst]: Oh, right, I'll fix that.

**Comment by androm3da** [comment]
Yeah, adding in the various LLVM vers is a good idea.  Can I deliver it in a separate submission?

File: README.rst [hunk 3]
```diff
@@ -26,9 +26,12 @@ versions of the `LLVM-C API`_ should be basically effortless.
 
 To try it out install LLVM and get `llvmcpy` using `pip`::
 
-    sudo apt-get install llvm
+    sudo apt-get install llvm libllvm3.8
```
Comment (aleclearmind) [inline README.rst]: Why do you want to force LLVM 3.8?

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# on travis, clang-3.5 in /usr/local includes an llvm-config
+#   that comes earlier in the path than the system llvm
+env:
+    - LLVM_CONFIG=/usr/bin/llvm-config
+
+before_install:
+    - dpkg -L libllvm3.5
+    - which llvm-config
+    - llvm-config --libdir
+    - rm -rf ~/.cache/
```
Comment (aleclearmind) [inline .travis.yml]: Why do we need these lines? Is there a situation where we might miss `llvm-config`? Can we reduce to a single line testing for `llvm-config` presence?

Also, I'm not familiar with travis, are runs stateful? Is `rm -rf ~/.cache/` necessary?

**Comment by aleclearmind** [comment]
Would you mind organize the PR as follows?

* A commit to fix py3k compatibility
* A commit for the `if len(libs) == 0:` thing
* A commit for `.travis.yml`, which for now I'd keep on a separate branch. If you want to integrate the multiple LLVM version thing later, it's OK for me. We can start with what you have right now.

Please try to organize the commit message with a first line shorter than 50 chars, a blank line and then some more lines (wrapped at 72 columns) explaining in further detail the commit, if necessary. Maybe you're already doing this, just making this explicit.

If it's too much effort I'll do it. ;)

File: README.rst (hunk 3 above)
Comment (androm3da) [inline README.rst]: There's no `libllvm` package, only specific-to-a-release ones.  I drafted other versions that explained this but it seemed too wordy.  But `libllvm` is necessary because it installs the `.so`s and `llvm` does not.

File: .travis.yml (hunk 4 above)
Comment (androm3da) [inline .travis.yml]: These were all debugging steps that are probably safe to purge.  I'll try cleaning up.

**Comment by androm3da** [comment]
I have no problem reorganizing it that way, sure.

AFAIK I can't do that with this PR, right?  I'll need to rebase and make a new PR.  If so, I'll close this one and get started on the new one.

**Comment by androm3da** [comment]
And if the commits are bound for separate branches, I think they have to be in different PRs.  I think I can keep it simple by having one commit for each submit.  

Is there a way for GH to squash or fixup the commits in the PR upon merge?  Otherwise I have to keep spawning new PRs if there's any edits required.

**Comment by aleclearmind** [comment]
I think there's a way to update the PR, try to dig a bit. Try to `push -f` to the branch associated to this PR.

**Comment by androm3da** [comment]
You're right, I found some suggestions that a force push will work.

File: README.rst (hunk 3 above)
Comment (androm3da) [inline README.rst]: I'll offer this `README` change as another PR.

**Comment by androm3da** [comment]
Ok I will leave this PR as-is and add the Travis CI content to another PR targeting another branch.

**Comment by aleclearmind** [comment]
Just one last thing, can you please rewrite the commit message from:

Evaluate 'args' iterator: Fix for py3k compat.

to:

py3k compatibility: materialize `args` iterator

The rest looks good and it's good to go.

**Comment by aleclearmind** [comment]
Merged, thanks!
